orchagent/vnetorch.cpp

```diff
@@ -0,0 +1,200 @@
+#include <unordered_map>
+#include <unordered_set>
+#include <exception>
+
+#include "sai.h"
+#include "macaddress.h"
+#include "orch.h"
+#include "request_parser.h"
+#include "vnetorch.h"
+
+extern sai_virtual_router_api_t* sai_virtual_router_api;
+extern sai_object_id_t gSwitchId;
+
+/*
+ * VRF Modeling and VNetVrf class definitions
+ */
+std::vector<VR_TYPE> vr_cntxt;
+
+bool VNetVrfObject::createObj(vector<sai_attribute_t>& attrs)
+{
+    auto l_fn = [&] (sai_object_id_t& router_id) {
+
+        sai_status_t status = sai_virtual_router_api->create_virtual_router(&router_id,
+                                                                            gSwitchId,
+                                                                            static_cast<uint32_t>(attrs.size()),
+                                                                            attrs.data());
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to create virtual router name: %s, rv: %d",
+                           vnet_name.c_str(), status);
+            throw std::runtime_error("Failed to create VR object");
+        }
+        return true;
+    };
+
+    /*
+     * Create ingress and egress VRF based on VR_VALID
+     */
+
+    for (auto vr_type : vr_cntxt)
+    {
+        sai_object_id_t router_id;
+        if (vr_type != VR_TYPE::VR_INVALID && l_fn(router_id))
+        {
+            vr_ids.insert(std::pair<VR_TYPE, sai_object_id_t>(vr_type, router_id));
+        }
+    }
+
+    SWSS_LOG_INFO("VNET '%s' router object created ", vnet_name.c_str());
+    return true;
+}
+
+bool VNetVrfObject::updateObj(vector<sai_attribute_t>& attrs)
+{
+    set<sai_object_id_t> vr_ent = getVRids();
+
+    for (const auto& attr: attrs)
+    {
+        for (auto it : vr_ent)
+        {
+            sai_status_t status = sai_virtual_router_api->set_virtual_router_attribute(it, &attr);
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("Failed to update virtual router attribute. VNET name: %s, rv: %d",
+                                vnet_name.c_str(), status);
+                return false;
+            }
+        }
+    }
+
+    SWSS_LOG_INFO("VNET '%s' was updated", vnet_name.c_str());
+    return true;
+}
+
+VNetVrfObject::~VNetVrfObject()
+{
+    set<sai_object_id_t> vr_ent = getVRids();
+    for (auto it : vr_ent)
+    {
+        sai_status_t status = sai_virtual_router_api->remove_virtual_router(it);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to remove virtual router name: %s, rv:%d",
+                            vnet_name.c_str(), status);
+        }
+    }
+
+    SWSS_LOG_INFO("VNET '%s' deleted ", vnet_name.c_str());
+}
+
+/*
+ * VNet Orch class definitions
+ */
+
+template <class T>
+std::unique_ptr<T> VNetOrch::createObject(const string& vnet_name, set<string>& plist,
+                                          vector<sai_attribute_t>& attrs)
```

Comment:
you make it a template,  are there class other than  VNetVrfObject using the same parameters of (const string&, set<string>&, vector<sai_attribute_t>&) ?

Reply:
yes. There is also a bridge model for Vnet. Hence the template.